### DISCUSSION TITLE
access to Glacier2 router in python

### DISCUSSION THREAD

**Discussion by FrankRulof** [zeroc-ice/ice#2065] [Q&A]
Hello,

I am writing a client for the mumble server and using Glacier2 and ICE as interfaces. I =created the following python code from a C++ piece of code the code is as follows 
import Ice
import Glacier2
import sys
import MumbleServer  # Your generated MumbleServer module

# Initialize the Ice communicator
initData = Ice.InitializationData()
initData.properties = Ice.createProperties()
initData.properties.setProperty('Ice.Default.Router', 'DemoGlacier2Router/router:default -p 4063 -h 192.168.1.63')
#initData.properties.setProperty('Ice.Default.Router', 'Glacier2/router:default -p 4063 -h 192.168.1.63')
iceCommunicator = Ice.initialize(initData)
print ("1")

try:
    # Get the default router from the communicator
    #murmurRouter = Glacier2.RouterPrx.uncheckedCast(iceCommunicator.getDefaultRouter())
    murmurRouter = Glacier2.RouterPrx.uncheckedCast(iceCommunicator.getDefaultRouter())
    print ("2")

    # Now you can use murmurRouter to create a session or perform other operations
    # For example:
    session = Glacier2.RouterPrx.createSession("SuperUser", "linuxtar")
    print ("session="+session)

    # Obtain a proxy for the Meta service
    meta = MumbleServer.MetaPrx.uncheckedCast(murmurRouter.getSession().ice_router)
    print ("meta="+meta)

when executing the code an error :: Error: RouterPrx.createSession() missing 1 required positional argument: 'password' I dont understand this error , the mumble server is running with the above user and password. The mumble client can connect. Can you help me with that ?
Thanks in advance

**Comment by pepone** [comment]
This is not correct:

```
session = Glacier2.RouterPrx.createSession("SuperUser", "linuxtar")
```

You must call `createSession` on the proxy you created before, something like:

```
session = murmurRouter.createSession("SuperUser", "linuxtar")
```

No clear what are you trying to do in the following code:
```
# Obtain a proxy for the Meta service
meta = MumbleServer.MetaPrx.uncheckedCast(murmurRouter.getSession().ice_router)
print ("meta="+meta)
```

You might want to take a look at the Python Glacier2 demo at https://github.com/zeroc-ice/ice-demos/blob/3.7/python/Glacier2/callback/client.py

**Comment by FrankRulof** [comment]
Thanks for your answer , i am going to look at the example you indicated 
To aswer your remark "No clear what are you trying to do in the following code: "The intend of the python code is to create a client that can control the administration (channels , ACL , users) of the mumble server, so meta is the base to be able to do so and i thought it needs the session from the router. can you guide me in that respect ?

**Comment by FrankRulof** [comment]
when i change "session = Glacier2.RouterPrx.createSession("SuperUser", "linuxtar")" into "session = murmurRouter.createSession("SuperUser", "linuxtar")" i get the error
Error: exception ::Ice::ObjectNotExistException
{
    id = 
    {
        name = router
        category = DemoGlacier2Router
    }
    facet = 
    operation = createSession
}
What does this error mean ? I hope you can help me

**Comment by pepone** [comment]
It means that the object with ID `DemoGlacier2Router/router` is not hosted by the target server.

A Glacier2  router hosts a well-known object whose identity is derived from the `Glacier2.InstanceName` configuration property. See https://doc.zeroc.com/ice/latest/ice-services/glacier2/getting-started-with-glacier2#id-.GettingStartedwithGlacier2v3.7-Glacier2ObjectIdentities

> No clear what are you trying to do in the following code: "The intend of the python code is to create a client that can control the administration (channels , ACL , users) of the mumble serve

It would be better to ask this question in the Mumble forum.

I just asked because the code doesn't look correct

```
# Obtain a proxy for the Meta service
meta = MumbleServer.MetaPrx.uncheckedCast(murmurRouter.getSession().ice_router)
print ("meta="+meta)
```

Here `murmurRouter` seems to be a proxy of type `Glacier2.RouterPrx` which doesn't have a `getSession` operation.
